core/src/main/java/io/grpc/internal/Http2ClientStreamTransportState.java

```diff
@@ -147,8 +147,13 @@ protected void transportDataReceived(ReadableBuffer frame, boolean endOfStream)
       inboundDataReceived(frame);
       if (endOfStream) {
         // This is a protocol violation as we expect to receive trailers.
-        transportError =
-            Status.INTERNAL.withDescription("Received unexpected EOS on DATA frame from server.");
+        if (frame.readableBytes() > 0) {
```

Comment:
The return value of `readableBytes()` may depend on the current read index, and `inboundDataReceived(frame)` may have changed the read index?

Reply:
Good catch. Fixed. This problem would have impacted #7277 as well.